config/secret.go

```diff
@@ -26,12 +24,48 @@ var secretStorePattern = regexp.MustCompile(`^\w+$`)
 // in a secret-store.
 var secretPattern = regexp.MustCompile(`@\{(\w+:\w+)\}`)
 
+// secretCount is the number of secrets use in Telegraf
 var secretCount atomic.Int64
 
+// selectedImpl is the configured implementation for secrets
+var selectedImpl secretImpl = &protectedSecretImpl{}
+
+// secretImpl represents an abstraction for different implementations of secrets
+type secretImpl interface {
+	Container(secret []byte) secretContainer
+	EmptyBuffer() SecretBuffer
+	Wipe(secret []byte)
+}
+
+// secretContainer represents an abstraction of the container holding the
+// actual secret value
+type secretContainer interface {
+	Destroy()
+	Equals(ref []byte) (bool, error)
+	Buffer() (SecretBuffer, error)
+	AsBuffer(secret []byte) SecretBuffer
+	Replace(secret []byte)
+}
+
+// SecretBuffer allows to access the content of the secret
+type SecretBuffer interface {
+	Size() int
+	Grow(capacity int)
+	Bytes() []byte
+	String() string
+	StringCopy() string
```

Comment:
I have re-written my response more times than I care to admit today. I keep going back and forth on what to actually call these. Rather, I figured I would ask questions:

Do we need both `String` and `StringCopy`? Instead should we have a single `String()` which does what your `StringCopy()` does, that way we have one canonical set of methods to use, and hence we avoid the confusion around the difference? 

As a reviewer the String method, as-is, is a foot gun asking for trouble and weird errors when you can not read the string anymore. I realize that is the intent, but how often do we really mean to do this?

Reply:
Thought I would share my other thoughts, specifically for naming:

* `String` --> `StringBuff`/`StringSecure`/`RawString`/`SecretString` 
* `StringCopy` --> `String`

Where the current `String`, which is the access to the secret buffer, gets renamed to specifically call out as the actually buffer string, and not just a generic string to aid the user and reviewers that this might disappear or is part of the secret store impl.

Reply:
I think we should keep both as we really want the current `String` because it allows to wipe the secret from memory. So how about the following `String -> TemporaryString` and `StringCopy -> String`?

Reply:
That works for me!